src/providers/postgres/qgspostgresdataitemguiprovider.cpp

```diff
@@ -735,3 +760,251 @@ void QgsPostgresDataItemGuiProvider::handleImportVector( QgsPGConnectionItem *co
+
+  const Qgis::ProjectFileFormat defaultProjectFileFormat = settings.enumValue( QStringLiteral( "/qgis/defaultProjectFileFormat" ), Qgis::ProjectFileFormat::Qgz );
+  const QString qgisProjectExt = tr( "QGIS Project Formats" ) + ( defaultProjectFileFormat == Qgis::ProjectFileFormat::Qgz ? " (*.qgz *.QGZ *.qgs *.QGS)" : " (*.qgs *.QGS *.qgz *.QGZ)" );
+  const QString qgzProjectExt = tr( "QGIS Bundled Project Format" ) + " (*.qgz *.QGZ)";
+  const QString qgsProjectExt = tr( "QGIS XML Project Format" ) + " (*.qgs *.QGS)";
+
+  QString filter;
+  const QString path = QFileDialog::getSaveFileName(
+    nullptr,
+    tr( "Save Project As" ),
+    defaultPath,
+    qgisProjectExt + QStringLiteral( ";;" ) + qgzProjectExt + QStringLiteral( ";;" ) + qgsProjectExt, &filter
+  );
+
+  if ( path.isEmpty() )
+    return;
+
+  QFileInfo fullPath( path );
+  QgsSettings().setValue( QStringLiteral( "UI/lastProjectDir" ), fullPath.path() );
+
+  const QString ext = fullPath.suffix().toLower();
+  if ( filter == qgisProjectExt && ext != QLatin1String( "qgz" ) && ext != QLatin1String( "qgs" ) )
+  {
+    switch ( defaultProjectFileFormat )
+    {
+      case Qgis::ProjectFileFormat::Qgs:
+      {
+        fullPath.setFile( fullPath.filePath() + ".qgs" );
+        break;
+      }
+      case Qgis::ProjectFileFormat::Qgz:
+      {
+        fullPath.setFile( fullPath.filePath() + ".qgz" );
+        break;
+      }
+    }
+  }
+  else if ( filter == qgzProjectExt && ext != QLatin1String( "qgz" ) )
+  {
+    fullPath.setFile( fullPath.filePath() + ".qgz" );
+  }
+  else if ( filter == qgsProjectExt && ext != QLatin1String( "qgs" ) )
+  {
+    fullPath.setFile( fullPath.filePath() + ".qgs" );
+  }
+
+  QgsProject project;
+  project.read( projectItem->path() );
+  project.setFileName( fullPath.filePath() );
+  const bool result = project.write();
+
+  if ( !result && context.messageBar() )
+  {
+    context.messageBar()->pushWarning( tr( "Export Project to File" ), tr( "Could not save project file" ) );
+  }
+}
+
+void QgsPostgresDataItemGuiProvider::renameProject( QgsPGProjectItem *projectItem, QgsDataItemGuiContext context )
+{
+  QgsNewNameDialog dlg( tr( "project “%1”" ).arg( projectItem->name() ) );
+  dlg.setWindowTitle( tr( "Rename Project" ) );
+  if ( dlg.exec() != QDialog::Accepted || dlg.name() == projectItem->name() )
+    return;
+
+  QgsPostgresConn *conn = QgsPostgresConn::connectDb( projectItem->postgresProjectUri().connInfo, false );
+  if ( !conn )
+  {
+    notify( tr( "Rename Project" ), tr( "Unable to rename project." ), context, Qgis::MessageLevel::Warning );
+    return;
+  }
+
+  const QString newUri = projectItem->uriWithNewName( dlg.name() );
+
+  // read the project, set title and new filename
+  QgsProject project;
+  project.read( projectItem->path() );
+  project.setTitle( dlg.name() );
+  project.setFileName( newUri );
+
+  if ( !QgsPostgresUtils::deleteProjectFromSchema( conn, projectItem->name(), projectItem->schemaName() ) )
```

Comment:
I'd flip the delete/write, so that we DON'T delete if the write fails. Otherwise we risk data loss.

Reply:
good point, thanks :)